### DISCUSSION TITLE
How to get the "buildPath" attribute of builds of a IProject

### DISCUSSION THREAD

**Discussion by thahnen** [eclipse-cdt/cdt#895] [Q&A]
Hello everyone,

## Context

I have quite a specific question, but first I'd like to give some context:

In our Eclipse plug-in, we list all the files in a project and want to exclude files that are not related to the actual project content - like the generated and compiled output. In case of Java it would be the output directory (`bin` / `build` / `target` / whatever).

For JDT projects it is quite simple (after checking that it is actually a JDT project):

```java
public Set<org.eclipse.core.runtime.IPath> getExclusions(org.eclipse.core.resources.IProject project) {
  var exclusions = new HashSet<IPath>();

  var javaProject = org.eclipse.jdt.core.JavaCore.create(project)
  if (javaProject == null) {
    return exclusions;
  }

  try {
    var outputLocation = javaProject.getOutputLocation();
    if (outputLocation != null) {
      exclusions.add(outputLocation);
    }

    for (var entry : javaProject.getResolvedClasspath(true)) {
      if (entry.getEntryKind() == org.eclipse.jdt.core.IClasspathEntry.CPE_SOURCE) {
        var output = entry.getOutputLocation();
        if (output != null) {
          exclusions.add(output);
        }
      }
    }
  } catch (JavaModelException ignored) {
    //
  }

  return exclusions;
}
```

This code snippet above is a bit simplified.

## Question

So, now leading to my queston. I'm not that familiar with CDT as well as CMake but from what I found out by testing by creating a normal C/C++ project inside Eclipse CPP as well as creating a simple CMake project and running the generator for *Eclipse CDT4 - Unix Makefiles* there doesn't seem to be something like a classpath.

I've checked the `.cproject` files from both projects and took a look at the CDT source code, starting with `org.eclipse.cdt.core.CCorePlugin`. There are this kind of `builder` tags that have a `buildPath` attribute in the XML.
But not for CMake projects that generated everything.

From the code, I couldn't find any reference except for it to be used in the `CCorePlugin#getExtensionScannerInfoProvider2` method. This seems to parse the XML structure?!

Questions:
 - **What is the best approach to get the `buildPath` attribute for the `builder` in a straight CDT project?**
 - **What is the approach for tools that generated the Eclipse files, like CMake for example?**

Thanks for some help and guidance in advance :bowing_man: 

Best,
Tobias

**Comment by jld01** [comment]
Hi @thahnen

For CDT _managed build_ projects, take a look at: [InternalCoreModelUtil.findBuildConfiguration()](https://github.com/eclipse-cdt/cdt/blob/main/core/org.eclipse.cdt.core/model/org/eclipse/cdt/internal/core/model/InternalCoreModelUtil.java#L101) for some inspiration.

The CDT [CMakeProjectGenerator](https://github.com/eclipse-cdt/cdt/blob/main/cmake/org.eclipse.cdt.cmake.core/src/org/eclipse/cdt/cmake/core/CMakeProjectGenerator.java#L64) assumes a build folder named `build/` under the project root.

Does this answer your question?

**Reply by thahnen** [reply]
HI @jld01,

Thanks for the fast response. I've taken a look at the `InternalCoreModelUtil.findBuildConfiguration()` for inspiration, which led me to the `ICConfigurationDescription#getBuildSettings()`, which in turn led me to `getResolvedOutputDirectories()` that looks very promising!

For the latter one, I'm not sure if this is the same generator as the one that I used that is "bundled" with CMake:

Inside the CMake project directory I ran `cmake -G"Eclipse CDT4 - Unix Makefiles" -D CMAKE_BUILD_TYPE=Debug .` which generated everything as expected but when imported in the IDE and building all the targets (no matter what), it generated the executable just inside the project folder and the object files under *CMakeFiles/CMakeProjectMain.dir*.
Maybe it is due to me running the command incorrectly, I just took it from the CMake website.

Looking at the `.project` file, it revealed that some arguments like `org.eclipse.cdt.make.core.build.location` and `org.eclipse.cdt.make.core.buildLocation` link both to my project directory directly.

It is also no big project, just one **CMakeLists.txt** and one **main.c**:

```cmake
set(CMAKE_C_COMPILER "/usr/bin/gcc")

PROJECT(CMakeProject
        LANGUAGES     C)

ADD_EXECUTABLE(CMakeProjectMain
               main.c)
```

```c
#include <stdio.h>
#include <stdlib.h>

int main(void) {
	puts("!!!Hello World!!!"); /* prints !!!Hello World!!! */
	return EXIT_SUCCESS;
}
```

In case I didn't run it correctly, I would have to access both these arguments somehow I guess and make sure that it is not the project directory itself - if the compiled code is stored there then so it is and we cannot exclude it?!

Best,
Tobias

**Comment by thahnen** [comment]
Just for everyone looking at this thinking about what the solution was: No direct solution was found for CMake, but an overall "good" solution was found:

```java
public Set<IPath> getExcludedPaths(IProject project) {
    var exclusions = new HashSet<IPath>();

    try {
      // 1) Check whether this is a CDT project
      var cProject = CoreModel.getDefault().getProjectDescription(project, false);
      if (cProject == null) {
        return exclusions;
      }
      var projectPath = project.getFullPath().makeAbsolute().toOSString();

      // 2) Iterate over all the different configurations, e.g. Debug or Release
      for (var config : cProject.getConfigurations()) {
        var configData = config.getConfigurationData();
        if (configData == null) {
          continue;
        }

        var buildData = configData.getBuildData();
        if (buildData == null) {
          continue;
        }

        // 3) Iterate over all the output directories as there can be multiple ones per configuration
        for (var outputDirectory : buildData.getOutputDirectories()) {
          var outputDirectoryPath = outputDirectory.getFullPath().makeAbsolute().toOSString();
          if (!projectPath.equals(outputDirectoryPath)) {
            // Only when the output directory is not the project directory we keep it. For example CMake projects can
            // generate CDT projects where the output is the project directory, this case should not be taken into
            // account and is up to the user. By default we will exclude compilation output when it is somewhere in a
            // sub-directory, e.g. "Debug" or "Release"!
            exclusions.add(org.eclipse.core.runtime.Path.fromOSString(
              "/" + project.getName() + "/" + outputDirectory.getFullPath()));
          }
        }
      }
    } catch (Exception err) {
      SonarLintLogger.get().error("Error while getting the exclusions of project '" + project.getName()
        + "' based on CDT!", err);
    }

    return exclusions;
}
```
